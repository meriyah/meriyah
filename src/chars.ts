/**
 * A list of character constants with much more human-readable names.
 */
export const enum Chars {

    Null = 0x00,

    Backspace      = 0x08,
    Tab            = 0x09,
    LineFeed       = 0x0A,
    VerticalTab    = 0x0B,
    FormFeed       = 0x0C,
    CarriageReturn = 0x0D,

    Space       = 0x20,
    Exclamation = 0x21,
    DoubleQuote = 0x22,
    Hash        = 0x23,
    Dollar      = 0x24,
    Percent     = 0x25,
    Ampersand   = 0x26,
    SingleQuote = 0x27,
    LeftParen   = 0x28,
    RightParen  = 0x29,
    Asterisk    = 0x2A,
    Plus        = 0x2B,
    Comma       = 0x2C,
    Hyphen      = 0x2D,
    Period      = 0x2E,
    Slash       = 0x2F,

    Zero         = 0x30,
    One          = 0x31,
    Two          = 0x32,
    Three        = 0x33,
    Four         = 0x34,
    Five         = 0x35,
    Six          = 0x36,
    Seven        = 0x37,
    Eight        = 0x38,
    Nine         = 0x39,
    Colon        = 0x3A,
    Semicolon    = 0x3B,
    LessThan     = 0x3C,
    EqualSign    = 0x3D,
    GreaterThan  = 0x3E,
    QuestionMark = 0x3F,

    UpperA = 0x41,
    UpperB = 0x42,
    UpperC = 0x43,
    UpperD = 0x44,
    UpperE = 0x45,
    UpperF = 0x46,
    UpperG = 0x47,
    UpperH = 0x48,
    UpperI = 0x49,
    UpperJ = 0x4A,
    UpperK = 0x4B,
    UpperL = 0x4C,
    UpperM = 0x4D,
    UpperN = 0x4E,
    UpperO = 0x4F,

    UpperP       = 0x50,
    UpperQ       = 0x51,
    UpperR       = 0x52,
    UpperS       = 0x53,
    UpperT       = 0x54,
    UpperU       = 0x55,
    UpperV       = 0x56,
    UpperW       = 0x57,
    UpperX       = 0x58,
    UpperY       = 0x59,
    UpperZ       = 0x5A,
    LeftBracket  = 0x5B,
    Backslash    = 0x5C,
    RightBracket = 0x5D,
    Caret        = 0x5E,
    Underscore   = 0x5F,

    Backtick = 0x60,
    LowerA   = 0x61,
    LowerB   = 0x62,
    LowerC   = 0x63,
    LowerD   = 0x64,
    LowerE   = 0x65,
    LowerF   = 0x66,
    LowerG   = 0x67,
    LowerH   = 0x68,
    LowerI   = 0x69,
    LowerJ   = 0x6A,
    LowerK   = 0x6B,
    LowerL   = 0x6C,
    LowerM   = 0x6D,
    LowerN   = 0x6E,
    LowerO   = 0x6F,

    LowerP      = 0x70,
    LowerQ      = 0x71,
    LowerR      = 0x72,
    LowerS      = 0x73,
    LowerT      = 0x74,
    LowerU      = 0x75,
    LowerV      = 0x76,
    LowerW      = 0x77,
    LowerX      = 0x78,
    LowerY      = 0x79,
    LowerZ      = 0x7A,
    LeftBrace   = 0x7B,
    VerticalBar = 0x7C,
    RightBrace  = 0x7D,
    Tilde       = 0x7E,
    Delete      = 0x7F,

    NextLine         = 0x85,
    NonBreakingSpace = 0xA0,

    Ogham = 0x1680,

    EnQuad             = 0x2000,
    EmQuad             = 0x2001,
    EnSpace            = 0x2002,
    EmSpace            = 0x2003,
    ThreePerEmSpace    = 0x2004,
    FourPerEmSpace     = 0x2005,
    SixPerEmSpace      = 0x2006,
    FigureSpace        = 0x2007,
    PunctuationSpace   = 0x2008,
    ThinSpace          = 0x2009,
    HairSpace          = 0x200A,
    ZeroWidthSpace     = 0x200B,
    ZeroWidthJoiner    = 0x200C,
    ZeroWidthNonJoiner = 0x200D,

    LineSeparator      = 0x2028,
    ParagraphSeparator = 0x2029,

    NarrowNoBreakSpace = 0x202F,
    MathematicalSpace  = 0x205F,
    IdeographicSpace   = 0x3000,

    ZeroWidthNoBreakSpace = 0xFEFF,

    ByteOrderMark = 0xFFEF,

    NonBMPMax = 0x10FFFF,

    LeadSurrogateMin = 0xD800,
    LeadSurrogateMax = 0xDBFF,
    TrailSurrogateMin = 0xDC00,
    TrailSurrogateMax = 0xDFFF,
    UTF16Max = 0xFFFF,
    NonBMPMin = 0x10000,
}
