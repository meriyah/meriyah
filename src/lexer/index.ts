export { scanSingleToken, nextToken } from './scan';
export { skipHashBang } from './comments';
export { advanceChar, consumeLineFeed, scanNewLine, LexerState, convertTokenType } from './common';
