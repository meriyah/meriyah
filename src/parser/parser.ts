import { convertTokenType } from '../lexer';
import { Token } from '../token';
import type * as ESTree from '../estree';
import {
  Flags,
  type OnComment,
  type OnInsertedSemicolon,
  type OnToken,
  type Location,
  type AssignmentKind,
  type DestructuringKind,
} from '../common';

export type ParserOptions = {
  shouldAddLoc?: boolean;
  shouldAddRanges?: boolean;
  /**
   * Used together with source maps. File containing the code being parsed
   */
  sourceFile?: string;
  /**
   * Holds either a function or array used on every comment
   */
  onComment?: OnComment;
  /**
   * Function invoked with the character offset when automatic semicolon insertion occurs
   */
  onInsertedSemicolon?: OnInsertedSemicolon;
  /**
   * Holds either a function or array used on every token
   */
  onToken?: OnToken;
};

export class Parser {
  private lastOnToken: [string, number, number, ESTree.SourceLocation] | null = null;

  token = Token.EOF;
  /**
   * The mutable parser flags, in case any flags need passed by reference.
   */
  flags = Flags.None;
  /**
   * The current index
   */
  index = 0;
  /**
   * Beginning of current line
   */
  line = 1;

  /**
   * Beginning of current column
   */
  column = 0;

  /**
   * Start position of whitespace/comment before current token
   */
  startIndex = 0;

  /**
   * The end of the source code
   */
  end = 0;

  /**
   * Start position of text of current token
   */
  tokenIndex = 0;

  /**
   * Start position of the column before newline
   */
  startColumn = 0;

  /**
   * Position in the input code of the first character after the last newline
   */
  tokenColumn = 0;

  /**
   * The number of newlines
   */
  tokenLine = 1;

  /**
   * Start position of text of current token
   */
  startLine = 1;

  /**
   * Holds the scanned token value
   */
  tokenValue: any = '';

  /**
   * Holds the raw text that have been scanned by the lexer
   */
  tokenRaw = '';

  /**
   * Holds the regExp info text that have been collected by the lexer
   */
  tokenRegExp: void | {
    pattern: string;
    flags: string;
  } = void 0;

  /**
   * The code point at the current index
   */
  currentChar = 0;

  /**
   *  https://tc39.es/ecma262/#sec-module-semantics-static-semantics-exportednames
   */
  exportedNames: Record<string, number> = {};

  /**
   * https://tc39.es/ecma262/#sec-exports-static-semantics-exportedbindings
   */

  exportedBindings: Record<string, number> = {};

  /**
   * Assignable state
   */
  assignable: AssignmentKind | DestructuringKind = 1;

  /**
   * Destructuring state
   */
  destructible: AssignmentKind | DestructuringKind = 0;

  /**
   * Holds leading decorators before "export" or "class" keywords
   */
  leadingDecorators: ESTree.Decorator[] = [];

  constructor(
    /**
     * The source code to be parsed
     */
    public readonly source: string,
    public readonly options: ParserOptions = {},
  ) {
    this.end = source.length;
    this.currentChar = source.charCodeAt(0);
  }

  /**
   * Get the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  getToken() {
    return this.token;
  }

  /**
   * Set the current token in the stream to consume
   * This function exists as workaround for TS issue
   * https://github.com/microsoft/TypeScript/issues/9998
   */
  setToken(value: Token, replaceLast = false) {
    this.token = value;

    const { onToken } = this.options;

    if (onToken) {
      if (value !== Token.EOF) {
        const loc = {
          start: {
            line: this.tokenLine,
            column: this.tokenColumn,
          },
          end: {
            line: this.line,
            column: this.column,
          },
        };

        if (!replaceLast && this.lastOnToken) {
          onToken(...this.lastOnToken);
        }
        this.lastOnToken = [convertTokenType(value), this.tokenIndex, this.index, loc];
      } else {
        if (this.lastOnToken) {
          onToken(...this.lastOnToken);
          this.lastOnToken = null;
        }
      }
    }
    return value;
  }

  get tokenStart(): Location {
    return {
      index: this.tokenIndex,
      line: this.tokenLine,
      column: this.tokenColumn,
    };
  }

  finishNode<T extends ESTree.Node>(node: T, start: Location, end: Location | void): T {
    if (this.options.shouldAddRanges) {
      node.start = start.index;
      const endIndex = end ? end.index : this.startIndex;
      node.end = endIndex;
      node.range = [start.index, endIndex];
    }

    if (this.options.shouldAddLoc) {
      node.loc = {
        start: {
          line: start.line,
          column: start.column,
        },
        end: end ? { line: end.line, column: end.column } : { line: this.startLine, column: this.startColumn },
      };

      if (this.options.sourceFile) {
        node.loc.source = this.options.sourceFile;
      }
    }

    return node;
  }
}

export function pushComment(comments: ESTree.Comment[], options: ParserOptions): OnComment {
  return function (type: ESTree.CommentType, value: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const comment: ESTree.Comment = {
      type,
      value,
    };

    if (options.shouldAddRanges) {
      comment.start = start;
      comment.end = end;
      comment.range = [start, end];
    }
    if (options.shouldAddLoc) {
      comment.loc = loc;
    }
    comments.push(comment);
  };
}

export function pushToken(tokens: any[], options: ParserOptions): OnToken {
  return function (type: string, start: number, end: number, loc: ESTree.SourceLocation) {
    const token: any = {
      token: type,
    };

    if (options.shouldAddRanges) {
      token.start = start;
      token.end = end;
      token.range = [start, end];
    }
    if (options.shouldAddLoc) {
      token.loc = loc;
    }
    tokens.push(token);
  };
}
