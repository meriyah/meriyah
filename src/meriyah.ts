import { Context } from './common';
import { parseSource, Options } from './parser';
import * as ESTree from './estree';
// Current version
import { version } from '../package.json';

/**
 * Parse a script, optionally with various options.
 */
export function parseScript(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

/**
 * Parse a module, optionally with various options.
 */
export function parseModule(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.Strict | Context.Module);
}

/**
 * Parse a module or a script, optionally with various options.
 */
export function parse(source: string, options?: Options): ESTree.Program {
  return parseSource(source, options, Context.None);
}

export { Options, ESTree, version };
