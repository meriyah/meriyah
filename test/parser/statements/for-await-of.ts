import * as t from 'node:assert/strict';
import { outdent } from 'outdent';
import { describe, it } from 'vitest';
import { parseSource } from '../../../src/parser';
import { fail } from '../../test-utils';

describe('Statements - For await of', () => {
  const wrappers = [
    { start: 'var a1 = async function*asyncGenWithName1(){', finish: '}' },
    { start: 'async function*asyncGenWithName2(){ ', finish: '}' },
    { start: 'async function asyncWithName2(){ ', finish: '}' },
    { start: 'class A { async * method() { ', finish: ' } }' },
    { start: 'var a1 = async () => {', finish: '}' },
    { start: 'var a1 = async () => { try {   ', finish: ' } catch (e) {} }' },
    { start: 'var a1 = async () => { {   ', finish: ' } }' },
    { start: 'var a1 = async () => { if (true) {   ', finish: ' } }' },
    { start: 'var a1 = async () => { if (true) ', finish: ' }' },
    {
      start: 'var a1 = async () => { if (true) foo(); else { ',
      finish: ' } }',
    },
    { start: 'var a1 = async () => { while (true) { ', finish: ' } }' },
    { start: 'var a1 = async () => { for(;;) { ', finish: ' } }' },
    { start: "var a1 = async () => { switch(e) { case '1' :  ", finish: ' } }' },
  ];

  const expressions = [
    'for await(const value of foo()) {}',
    'for await(let value of foo()) {}',
    'for await(var value of foo()) {}',
    'for await(var [a, b] of foo()) {}',
    'for await(let {a, b} of foo()) {}',
    'for await(const [... b] of foo()) {}',
    'for await(const [,,, b] of foo()) {}',
    'for await(const value of boo) {}',
    'for await(let value of boo) {}',
    'for await(const value of foo().boo()) {}',
    'for await(let value of foo.boo()) {}',
    'for await(let value of foo.boo(value)) {}',
    'for await(let value of [1,2,3]) {}',
    'for await(value of [1,2,3]) {}',
    'for await(value of x + x) {}',
    'for await(value of f()) {}',
    'for await(value of (x + x)) {}',
  ];

  wrappers.forEach((wrapper) => {
    expressions.forEach((exp) => {
      it(wrapper.start + exp + wrapper.finish, () => {
        t.doesNotThrow(() => {
          parseSource(wrapper.start + exp + wrapper.finish);
        });
      });
      it(wrapper.start + exp + wrapper.finish, () => {
        t.doesNotThrow(() => {
          parseSource(wrapper.start + exp + wrapper.finish, { lexical: true });
        });
      });
    });
  });

  for (const arg of [
    'for await (x of []) function d() {};',
    'for await (x of []) function d() {}; return d;',
    'for await (x of []) function* g() {};',
    'for await (x of []) function* g() {}; return g;',
    'for await (x of []) async function a() {};',
    'for await (x of []) async function a() {}; return a;',
  ]) {
    it(`async function f() {${arg} }`, () => {
      t.throws(() => {
        parseSource(`async function f() { ${arg} }`);
      });
    });

    it(`async function f() {${arg} }`, () => {
      t.throws(() => {
        parseSource(`async function f() { ${arg} }`, { webcompat: true });
      });
    });

    it(`async function f() { 'use strict'; ${arg} }`, () => {
      t.throws(() => {
        parseSource(`async function f() { 'use strict'; ${arg} }`);
      });
    });
  }

  for (const arg of [
    '(a = 1 of [])',
    '(a = 1) of [])',
    '(a.b = 1 of [])',
    '([a] = 1 of [])',
    '(([a] = 1) of [])',
    '([a = 1] = 1 of [])',
    '(([a = 1] = 1) of [])',
    '([a = 1 = 1, ...b] = 1 of [])',
    '(([a = 1 = 1, ...b] = 1) of [])',
    '({a} = 1 of [])',
    '(({a} = 1) of [])',
    '({a: a} = 1 of [])',
    '(({a: a} = 1) of [])',
    "({'a': a} = 1 of [])",
    "(({'a': a} = 1) of [])",
    '({"a": a} = 1 of [])',
    '(({"a": a} = 1) of [])',
    '({[Symbol.iterator]: a} = 1 of [])',
    '(({[Symbol.iterator]: a} = 1) of [])',
    '({0: a} = 1 of [])',
    '({0: a = 1} = 1 of [])',
    '(({0: a = 1} = 1) of [])',
    '(function a() {} of [])',
    '([1] of [])',
    '({a: 1} of [])',
    '(var a = 1 of [])',
    '(var a, b of [])',
    '(var [a] = 1 of [])',
    '(var [a], b of [])',
    '(var [a = 1] = 1 of [])',
    '(var [a = 1], b of [])',
    '(var [a = 1 = 1, ...b] of [])',
    '(var [a = 1, ...b], c of [])',
    '(var {a} = 1 of [])',
    '(var {a}, b of [])',
    '(var {a: a} = 1 of [])',
    '(var {a: a}, b of [])',
    "(var {'a': a} = 1 of [])",
    "(var {'a': a}, b of [])",
    '(var {"a": a} = 1 of [])',
    '(var {"a": a}, b of [])',
    '(var {[Symbol.iterator]: a} = 1 of [])',
    '(var {[Symbol.iterator]: a}, b of [])',
    '(var {0: a} = 1 of [])',
    '(var {0: a}, b of [])',
    '(var {a = 1} = 1 of [])',
    '(var {a = 1}, b of [])',
    '(var {a: a = 1} = 1 of [])',
    '(var {a: a = 1}, b of [])',
    "(var {'a': a = 1} = 1 of [])",
    "(var {'a': a = 1}, b of [])",
    '(var {"a": a = 1} = 1 of [])',
    '(var {"a": a = 1}, b of [])',
    '(var {[Symbol.iterator]: a = 1} = 1 of [])',
    '(var {[Symbol.iterator]: a = 1}, b of [])',
    '(var {0: a = 1} = 1 of [])',
    '(var {0: a = 1}, b of [])',
    '(let a = 1 of [])',
    '(let a, b of [])',
    '(let [a] = 1 of [])',
    '(let [a], b of [])',
    '(let [a = 1] = 1 of [])',
    '(let [a = 1], b of [])',
    '(let [a = 1, ...b] = 1 of [])',
    '(let [a = 1, ...b], c of [])',
    '(let {a} = 1 of [])',
    '(let {a}, b of [])',
    '(let {a: a} = 1 of [])',
    '(let {a: a}, b of [])',
    "(let {'a': a} = 1 of [])",
    "(let {'a': a}, b of [])",
    '(let {"a": a} = 1 of [])',
    '(let {"a": a}, b of [])',
    '(let {[Symbol.iterator]: a} = 1 of [])',
    '(let {[Symbol.iterator]: a}, b of [])',
    '(let {0: a} = 1 of [])',
    '(let {0: a}, b of [])',
    '(let {a = 1} = 1 of [])',
    '(let {a = 1}, b of [])',
    '(let {a: a = 1} = 1 of [])',
    '(let {a: a = 1}, b of [])',
    "(let {'a': a = 1} = 1 of [])",
    "(let {'a': a = 1}, b of [])",
    '(let {"a": a = 1} = 1 of [])',
    '(let {"a": a = 1}, b of [])',
    '(let {[Symbol.iterator]: a = 1} = 1 of [])',
    '(let {[Symbol.iterator]: a = 1}, b of [])',
    '(let {0: a = 1} = 1 of [])',
    '(let {0: a = 1}, b of [])',
    '(const a = 1 of [])',
    '(const a, b of [])',
    '(const [a] = 1 of [])',
    '(const [a], b of [])',
    '(const [a = 1] = 1 of [])',
    '(const [a = 1], b of [])',
    '(const [a = 1, ...b] = 1 of [])',
    '(const [a = 1, ...b], b of [])',
    '(const {a} = 1 of [])',
    '(const {a}, b of [])',
    '(const {a: a} = 1 of [])',
    '(const {a: a}, b of [])',
    "(const {'a': a} = 1 of [])",
    "(const {'a': a}, b of [])",
    '(const {"a": a} = 1 of [])',
    '(const {"a": a}, b of [])',
    '(const {[Symbol.iterator]: a} = 1 of [])',
    '(const {[Symbol.iterator]: a}, b of [])',
    '(const {0: a} = 1 of [])',
    '(const {0: a}, b of [])',
    '(const {a = 1} = 1 of [])',
    '(const {a = 1}, b of [])',
    '(const {a: a = 1} = 1 of [])',
    '(const {a: a = 1}, b of [])',
    "(const {'a': a = 1} = 1 of [])",
    "(const {'a': a = 1}, b of [])",
    '(const {"a": a = 1} = 1 of [])',
    '(const {"a": a = 1}, b of [])',
    '(const {[Symbol.iterator]: a = 1} = 1 of [])',
    '(const {[Symbol.iterator]: a = 1}, b of [])',
    '(const {0: a = 1} = 1 of [])',
    '(const {0: a = 1}, b of [])',
    '(;;)',
    '(x;y;z)',
    '(const x in y)',
    '(let x in y)',
    '(var x in y)',
    '([x] in y)',
    '({x} in y)',
    '("foo".x in y)',
    'for await (a in b) {}',
    'for await (;;) {}',
  ]) {
    it(`async function f() { for await ${arg}; }`, () => {
      t.throws(() => {
        parseSource(`async function f() { for await ${arg}; }`);
      });
    });
    it(`async function f() { for await ${arg}; }`, () => {
      t.throws(() => {
        parseSource(`async function f() { for await ${arg}; }`, { lexical: true });
      });
    });
    it(`async function f() { for await ${arg}; }`, () => {
      t.throws(() => {
        parseSource(`async function f() { for await ${arg}; }`, { webcompat: true, lexical: true });
      });
    });
    it(`async function f() { 'use strict'; for await ${arg}; }`, () => {
      t.throws(() => {
        parseSource(`async function f() { 'use strict'; for await ${arg}; }`);
      });
    });

    it(`async function f() { for await  ${arg}  {} }`, () => {
      t.throws(() => {
        parseSource(`async function f() { for await  ${arg} {} }`);
      });
    });

    it(`async function * f() { 'use strict'; for await ${arg}; }`, () => {
      t.throws(() => {
        parseSource(`async function * f() { 'use strict'; for await ${arg}; }`);
      });
    });

    it(`async function * f() { for await ${arg}; }`, () => {
      t.throws(() => {
        parseSource(`async function * f() { for await ${arg}; }`);
      });
    });
  }

  for (const arg of [
    '(a of [])',
    '(a.b of [])',
    '([a] of [])',
    '([a = 1] of [])',
    '([a = 1, ...b] of [])',
    '({a} of [])',
    '({a: a} of [])',
    "({'a': a} of [])",
    '({"a": a} of [])',
    '({[Symbol.iterator]: a} of [])',
    '({0: a} of [])',
    '({a = 1} of [])',
    '({a: a = 1} of [])',
    "({'a': a = 1} of [])",
    '({"a": a = 1} of [])',
    '({[Symbol.iterator]: a = 1} of [])',
    '({0: a = 1} of [])',
    '(var x of [])',
    '(var [a] of [])',
    '(var [a = 1] of [])',
    '(var [a = 1, ...b] of [])',
    '(var {a} of [])',
    '(var {a: a} of [])',
    "(var {'a': a} of [])",
    '(var {"a": a} of [])',
    '(var {[Symbol.iterator]: a} of [])',
    '(var {0: a} of [])',
    '(var {a = 1} of [])',
    '(var {a: a = 1} of [])',
    "(var {'a': a = 1} of [])",
    '(var {"a": a = 1} of [])',
    '(var {[Symbol.iterator]: a = 1} of [])',
    '(var {0: a = 1} of [])',
    '(let a of [])',
    '(let [a] of [])',
    '(let [a = 1] of [])',
    '(let [a = 1, ...b] of [])',
    '(let {a} of [])',
    '(let {a: a} of [])',
    "(let {'a': a} of [])",
    '(let {"a": a} of [])',
    '(let {[Symbol.iterator]: a} of [])',
    '(let {0: a} of [])',
    '(let {a = 1} of [])',
    '(let {a: a = 1} of [])',
    "(let {'a': a = 1} of [])",
    '(let {"a": a = 1} of [])',
    '(let {[Symbol.iterator]: a = 1} of [])',
    '(let {0: a = 1} of [])',
    '(/foo/g[x] of c)',
    '(/foo/g.x of c)',
    '(const a of [])',
    '(const [a] of [])',
    '(const [a = 1] of [])',
    '(const [a = 1, ...b] of [])',
    '(const {a} of [])',
    '(const {a: a} of [])',
    "(const {'a': a} of [])",
    '(const {"a": a} of [])',
    '(const {[Symbol.iterator]: a} of [])',
    '(const {0: a} of [])',
    '(const {a = 1} of [])',
    '(const {a: a = 1} of [])',
    "(const {'a': a = 1} of [])",
    '(const {"a": a = 1} of [])',
    '(const {[Symbol.iterator]: a = 1} of [])',
    '(const {0: a = 1} of [])',
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function f() { let y; for await  ${arg}; }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function * f() { 'use strict'; for await\n ${arg}; }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function f() { 'use strict'; for await ${arg}  { } }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function f() { 'use strict'; for await ${arg}  { } }`, { webcompat: true });
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function * f() { 'use strict'; for await  ${arg}  { } }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function f() { for\nawait  ${arg}  { } }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function f() { 'use strict'; for\nawait  ${arg}  { } }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function * f() { for await ${arg}; }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function * f() { for await ${arg}; }`, { webcompat: true });
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function f() { let y; for await  ${arg}; }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function f() { 'use strict'; let y; for\nawait ${arg}{ } }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function f() { 'use strict'; let y; for\nawait ${arg}; }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function f() { 'use strict'; let y; for await ${arg}; }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function * f() { 'use strict'; let y; for await\n ${arg}; }`);
      });
    });
  }

  fail('Statements - For await of (fail)', [
    {
      code: 'async function fn() { for await (var [...[ x ] = []] of (async function*() { yield* [[]]; })()) {}}',
      options: { impliedStrict: true },
    },
  ]);

  for (const arg of [
    'for await ([v2 = 10, vNull = 11, vHole = 12, vUndefined = 13, vOob = 14] of [[2, null, , undefined]]) {}',
    'for await ([ a = x += 1, b = x *= 2 ] of [[]]) {}',
    'for await ([ a = x += 1, b = x *= 2 ] of [[]]) {}',
    'for await ([[ _ ]] of [[ , ]]) {}',
    'for await ([[ x ]] of [[]]) {}',
    'for await ([{ x }] of [[null]]) {}',
    'for await ([{ x }] of [[ , ]]) {}',
    'for await ([{ x = yield }] of [[{}]]) {}',
    'for await ([{ x }] of [[{ x: 2 }]]) {}',
    outdent`
      for await ([] of ["foo"
      ]) {}
    `,
    'for await ({ xCls = class x {}, cls = class {}, xCls2 = class { static name() {} } } of [{}]) {}',
    'for await ([...{ 0: x, length }] of [[]]) {}',
    'for await ({} of [false]) {}',
    'for await ({ w, x, y } of [{ x: 5 }]) {}',
    'for await ({ xCover = (0, function() {}), cover = (function() {}) } of [{}]) {}',
    'for await ({ eval = 3, arguments = 4 } of [{}]) {}',
    'for await ({ unresolvable } of [{}]) {}',
    'for await ({ eval = 3, arguments = 4 } of [{}]) {}',
    'for await ({ x = yield } of [{}]) {}',
    'for await ({ y: x = 1 } of [{ y: null }]) {}',
    'for await ({ x: x[yield] } of [{ x: 23 }]) {}',
    'for await ({ x: { x = yield } } of [{ x: {} }]) {}',
    'for await ({...src.y} of [{ x: 1, y: 2}]) {}',
    'for await (const [x] of [iter]) {}',
    'for await (const [[x]] of [[null]]) {}',
    'for await (const [x = 23] of [[,]]) {}',
    'for await (const [{ x, y, z } = { x: 44, y: 55, z: 66 }] of [[{ x: 11, y: 22, z: 33 }]]) {}',
    'for await (const [...[,]] of [g()]) {}',
    'for await (const [...x] of [[1, 2, 3]]) {}',
    'for await (const [x] of (async function*() { yield* [[]]; })()) {}',
    'for await (const [{ u: v, w: x, y: z } = { u: 444, w: 555, y: 666 }] of (async function*() { yield* [[{ u: 777, w: 888, y: 999 }]]; })()) {}',
    'for await (const { cover = (function () {}), xCover = (0, function() {})  } of (async function*() { yield* [{}]; })()) {}',
    'for await (const { w: [x, y, z] = [4, 5, 6] } of (async function*() { yield* [{ w: [7, undefined, ] }]; })()) {}',
    'for await (let [[...x] = function() {}()] of [[[2, 1, 3]]]) {}',
    'for await (let [x = 23] of [[]]) {}',
    'for await (let [...[]] of [function*() {}()]) {}',
    outdent`
      for await (let [x] of (async function*() {
        yield* [{}];
      })()) {}
    `,
    'for await (let [x = 23] of (async function*() { yield* [[undefined]]; })()) {}',
    'for await (let [{ x, y, z } = { x: 44, y: 55, z: 66 }] of (async function*() { yield* [[{ x: 11, y: 22, z: 33 }]]; })()) {}',
    'for await (let { w: { x, y, z } = { x: 4, y: 5, z: 6 } } of [{ w: { x: undefined, z: 7 } }]) {}',
    'for await (var [cls = class {}, xCls = class X {}, xCls2 = class { static name() {} }] of [[]]) {}',
    'for await (var [x] of [[]]) {}',
    'for await (var { w: { x, y, z } = undefined } of [{ }]) { return; }',
    'for await ([ xGen = function* x() {}, gen = function*() {} ] of [[]]) {}',
    'for await ({} of [false]) {}',
    'for await ({ x, y } of [{ x: 3 }]) {}',
    'for await ({ xCover = (0, function() {}), cover = (function() {}) } of [{}]) {}',
    'for await ({ y: x = 1 } of [{ y: 2 }]) {}',
    'for await ({ x: { x = yield } } of [{ x: {} }]) {}',
    'for await (const [cls = class {}, xCls = class X {}, xCls2 = class { static name() {} }] of [[]]) {}',
    'for await (const [{ x }] of [[null]]) {}',
    'for await (const [...{ 0: v, 1: w, 2: x, 3: y, length: z }] of [[7, 8, 9]]) {}',
    'for await (const {} of [undefined]) {}',
    'for await (x of y) {}',
    'for await ([x] of y) {}',
    'for await ({x} of y) {}',
    'for await ("foo".x of y) {}',
    'for await (var x of y) {}',
    'for await (var x of y) {}',
    'for ([].x in y);',
    'for await (let x of y) {}',
    'for await (const x of y) {}',
    'for await ((x) of y) {}',
    'for await (let [[x, y, z] = [4, 5, 6]] of [[]]) {}',
    'for await (let [[] = function() {}()] of [[]]) {}',
    'for await (let [cls = class {}, xCls = class X {}, xCls2 = class { static name() {} }] of [[]]) {}',
    'for await (let [fn = function () {}, xFn = function x() {}] of [[]]) {}',
    'for await (const [{ x, y, z } = { x: 44, y: 55, z: 66 }] of (async function*() { yield* [[]]; })()) {}',
    'for await (const [{ u: v, w: x, y: z } = { u: 444, w: 555, y: 666 }] of (async function*() { yield* [[{ u: 777, w: 888, y: 999 }]];})()) {}',
    'for await (const [...x] of [function*() {}()]) {}',
    'for (var { w: { x, y, z } = { x: 4, y: 5, z: 6 } } of [{ w: null }]) {}',
    'for ([x] of [[0]]) {}',
    'for await (a of b) {}',
    'for await ("foo"[x] of c) d;',
    'for await (456..x of c) d;',
    'for await (456[x] of c) d;',
    'for await (((x)=>{}).x of y);',
    'for await (let a of b) {} for (let c of d) {}',
    'for await ([...[x[yield]]] of [[86]]) {}',
    'for await ([...{ 0: x, length }] of [[null]]) {}',
    'for (var { w: [x, y, z] = [4, 5, 6] } of [{ w: null }]) {}',
    'for await (const { x, } of [{ x: 23 }]) {}',
    'for await (const { w: { x, y, z } = { x: 4, y: 5, z: 6 } } of [{ w: { x: undefined, z: 7 } }]) {}',
    'for await (let [{ x, y, z } = { x: 44, y: 55, z: 66 }] of (async function*() { yield* [[{ x: 11, y: 22, z: 33 }]]; })()) {}',
  ]) {
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function fn() { ${arg} }`);
      });
    });
    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async function *fn() { ${arg} }`);
      });
    });

    it(`${arg}`, () => {
      t.doesNotThrow(() => {
        parseSource(`async () => { ${arg} }`);
      });
    });
  }

  it('accepts top level for await in module context', () => {
    t.doesNotThrow(() => {
      parseSource('for await (const a of b) {}', { sourceType: 'module' });
    });

    t.throws(() => {
      parseSource('for await (const a of b) {}');
    });

    t.throws(() => {
      parseSource('function c() { for await (const a of b) {} }', { sourceType: 'module' });
    });
  });
});
