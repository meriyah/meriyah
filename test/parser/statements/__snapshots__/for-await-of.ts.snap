// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Statements - For await of > Statements - For await of (fail) > async function fn() { for await (var [...[ x ] = []] of (async function*() { yield* [[]]; })()) {}} 1`] = `
"SyntaxError [1:53-1:55]: The binding declaration is not destructible
> 1 | async function fn() { for await (var [...[ x ] = []] of (async function*() { yield* [[]]; })()) {}}
    |                                                      ^^ The binding declaration is not destructible"
`;
